src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -0,0 +1,762 @@
+                int topBlockShift = 64 - bottomBlockShift;
+                int middleBlockShift = topBlockShift - 32;
+
+                exponent += (int)(topBlockBits);
+
+                uint bottomBlock = value.GetBlock(bottomBlockIndex);
+                uint bottomBits = bottomBlock >> bottomBlockShift;
+
+                ulong middleBits = (ulong)(value.GetBlock(middleBlockIndex)) << middleBlockShift;
+                ulong topBits = (ulong)(value.GetBlock(topBlockIndex)) << topBlockShift;
+
+                mantissa = topBits + middleBits + bottomBits;
+
+                uint unusedBottomBlockBitsMask = (1u << (int)(topBlockBits)) - 1;
+                hasZeroTail &= (bottomBlock & unusedBottomBlockBitsMask) == 0;
+            }
+
+            for (uint i = 0; i != bottomBlockIndex; i++)
+            {
+                hasZeroTail &= (value.GetBlock(i) == 0);
+            }
+
+            return AssembleFloatingPointBits(in info, mantissa, exponent, hasZeroTail);
+        }
+
+        // get 32-bit integer from at most 9 digits
+        private static uint DigitsToUInt32(char* p, int count)
+        {
+            Debug.Assert((1 <= count) && (count <= 9));
+
+            char* end = (p + count);
+            uint res = (uint)(p[0] - '0');
+
+            for (p++; p < end; p++)
+            {
+                res = (10 * res) + p[0] - '0';
+            }
+
+            return res;
+        }
+
+        // get 64-bit integer from at most 19 digits
+        private static ulong DigitsToUInt64(char* p, int count)
+        {
+            Debug.Assert((1 <= count) && (count <= 19));
+
+            char* end = (p + count);
+            ulong res = (ulong)(p[0] - '0');
+
+            for (p++; p < end; p++)
+            {
+                res = (10 * res) + p[0] - '0';
+            }
+
+            return res;
+        }
+
+        // Get the 64-bit product of two 32-bit integers
+        private static ulong Multiply32x32To64(uint lhs, uint rhs)
+        {
+            return (ulong)(lhs) * rhs;
+        }
+
+        // Get the high 64-bits of the product of two 64-bit integers
+        private static ulong Multiply64x64To128(ulong lhs, ulong rhs, out ulong lower)
+        {
+            uint al = (uint)(lhs);
+            uint au = (uint)(lhs >> 32);
+
+            uint bl = (uint)(rhs);
+            uint bu = (uint)(rhs >> 32);
+
+            ulong c = Multiply32x32To64(al, bl);
+            uint cl = (uint)(c);
+            uint cu = (uint)(c >> 32);
+
+            ulong d = Multiply32x32To64(au, bl) + cu;
+            uint dl = (uint)(d);
+            uint du = (uint)(d >> 32);
+
+            ulong e = Multiply32x32To64(al, bu) + dl;
+            uint eu = (uint)(e >> 32);
+
+            lower = (e << 32) + eu;
+            return Multiply32x32To64(au, bu) + du + eu;
+        }
+
+        private static ulong NumberToFloatingPointBits(ref NumberBuffer number, in FloatingPointInfo info)
```

Comment:
This method is ~300 lines long now - I think it is close to hitting the various internal JIT limits, if it is not over them already. Would it make sense to split the slow path into a separate method at least?

Reply:
Yes, I'll split it.